### PR TITLE
Deprecate: Remove Python2.7 installation support

### PR DESCRIPTION
**Pull request by jfuss**
*Issue #, if available:*
#1273

*Description of changes:*
Removes the ability to install SAM CLI into a Py2 environment.

I want to take this time to also upgrade some of our dev dependencies that were stuck in the stone age and get rid of six. We will need to do similar things in `aws_lambda_builders` but that will be completely separate than this. If some of the dev dependency upgrades become to much, I will move that to a different PR but for now keeping this as a Draft PR.

*Checklist:*

- [ ] Write Design Document ([Do I need to write a design document?](https://github.com/awslabs/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.rst#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [ ] `make pr` passes
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by TheSriram** [comment]
Nice 🔥🔥, next up all of the six stuff, six.moves etc.

File: requirements/dev.txt [hunk 1]
```diff
@@ -1,17 +1,9 @@
...
-mock==2.0.0
-parameterized==0.6.1
-pathlib2==2.3.2; python_version<"3.4"
-futures==3.2.0; python_version<"3.2.3"
-# Py3.2 backport
-backports.tempfile==1.0
-pytest-xdist==1.20.0
\ No newline at end of file
+pytest==5.1.2
+#mock==2.0.0 # mock is now part of the Python standard library, available as unittest.mock in Python 3.3 onwards.
```
Comment (TheSriram) [inline requirements/dev.txt]: Should we remove this completely?

File: samcli/commands/local/cli_common/user_exceptions.py [hunk 2]
```diff
@@ -10,68 +10,50 @@ class InvokeContextException(UserException):
...
 
 class InvalidLayerVersionArn(UserException):
     """
     The LayerVersion Arn given in the template is Invalid
     """
 
-    pass
-
 
 class UnsupportedIntrinsic(UserException):
```
Comment (TheSriram) [inline samcli/commands/local/cli_common/user_exceptions.py]: We dont need a pass anymore? Didnt know this. Good to know.

File: samcli/commands/local/lib/swagger/reader.py [hunk 3]
```diff
@@ -152,7 +154,7 @@ def _download_swagger(self, location):
         """
 
         if not location:
-            return
```
Comment (TheSriram) [inline samcli/commands/local/lib/swagger/reader.py]: Are we just being explicit? or is this the norm with Python3?

File: samcli/lib/build/app_builder.py [hunk 4]
```diff
@@ -322,8 +318,7 @@ def _parse_builder_response(stdout_data, image_name):
                 LOG.debug("Builder library does not support the supplied method")
                 raise UnsupportedBuilderLibraryVersionError(image_name, msg)
 
-            else:
-                LOG.debug("Builder crashed")
-                raise ValueError(msg)
+            LOG.debug("Builder crashed")
```
Comment (TheSriram) [inline samcli/lib/build/app_builder.py]: Not blocking on this, but we have the same verbiage on the 'builder crashed' on if there was an exception or if there was error code we dont understand. We need to differentiate on that.

File: tests/smoke/templates/sar/AmazonConnectSalesforceLambda-template.yaml [hunk 5]
```diff
@@ -1,3 +1,4 @@
+# -*- coding: utf-8 -*-
```
Comment (TheSriram) [inline tests/smoke/templates/sar/AmazonConnectSalesforceLambda-template.yaml]: probably not needed.

File: requirements/dev.txt (hunk 1 above)
Comment (jfuss) [reply requirements/dev.txt]: O yes this should be.

File: samcli/commands/local/cli_common/user_exceptions.py (hunk 2 above)
Comment (jfuss) [reply samcli/commands/local/cli_common/user_exceptions.py]: https://github.com/PyCQA/pylint/issues/2208

File: samcli/commands/local/lib/swagger/reader.py (hunk 3 above)
Comment (jfuss) [reply samcli/commands/local/lib/swagger/reader.py]: This falls under pylint's inconsistent-return-statements (R1710), see http://pylint.pycqa.org/en/latest/technical_reference/features.html for more details.

I ended up just updating everything to make this go away. We could instead ignore this and revert the commit with all these types of changes.

**Comment by TheSriram** [comment]
Also could you remove: https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/cli/command.py#L27 as well :)

File: tests/smoke/templates/sar/AmazonConnectSalesforceLambda-template.yaml (hunk 5 above)
Comment (jfuss) [reply tests/smoke/templates/sar/AmazonConnectSalesforceLambda-template.yaml]: Yes I removed that but forgot to commit it to see if windows would pass.

**Comment by jfuss** [comment]
> Also could you remove: https://github.com/awslabs/aws-sam-cli/blob/develop/samcli/cli/command.py#L27 as well :)

Good call.

File: samcli/lib/utils/osutils.py [hunk 6]
```diff
@@ -49,16 +49,7 @@ def stdout():
...
-    # We write all of the data to stdout with bytes, typically io.BytesIO. stdout in Python2
-    # accepts bytes but Python3 does not. This is due to a type change on the attribute. To keep
-    # this consistent, we leave Python2 the same and get the .buffer attribute on stdout in Python3
-    byte_stdout = sys.stdout
-
-    if sys.version_info.major > 2:
-        byte_stdout = sys.stdout.buffer  # pylint: disable=no-member
-
-    return byte_stdout
+    return sys.stdout.buffer
```
Comment (sanathkr) [inline samcli/lib/utils/osutils.py]: Yaay!

**Pull request merged**